### PR TITLE
NSE: Preview support

### PR DESCRIPTION
**Pull request by gokmen**
This PR provides dynamic Preview support for Stack editor in NSE and also preview support for Readme pane.

Bonus: `JStackTemplate.preview` endpoint will provide a preview of provided stack template content.

### CODE REVIEW

File: client/app/lib/util/stacks/generatepreview.coffee [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+      for field in data
+
+        if type is 'userInput'
+          warnings.userInput ?= []
+          warnings.userInput.push field
+          continue
+
+        if content = jspath.getAt availableData[type], field
+          search   = if type is 'custom'  \
```
Comment (sinan) [inline client/app/lib/util/stacks/generatepreview.coffee]: no need for `\` and the indentation below

File: client/app/lib/util/stacks/generatepreview.coffee [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            then ///\${var.#{type}_#{field}}///g
```
Comment (sinan) [inline client/app/lib/util/stacks/generatepreview.coffee]: `///\${var\.#{type}_#{field}}///g` escape the `.`

File: client/app/lib/util/stacks/generatepreview.coffee [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            else ///\${var.koding_#{type}_#{field}}///g
```
Comment (sinan) [inline client/app/lib/util/stacks/generatepreview.coffee]: again the `\.`

File: client/new-stack-editor/lib/views/editor.coffee [hunk 4]
```diff
@@ -15,23 +17,41 @@ module.exports = class Editor extends BaseView
...
     @aceView = new AceView {
       cssClass: 'editor'
       useStorage: no
       createBottomBar: no
     }, FSHelper.createFileInstance { path: "localfile:/#{filename}" }
 
+    @getPreview = preview  if preview
+
+
+  getContent: (as = 'is') ->
```
Comment (sinan) [inline client/new-stack-editor/lib/views/editor.coffee]: minor: i dont think the default param and the early return is necessary

File: website/swagger.json [hunk 5]
```diff
@@ -5460,6 +5460,77 @@
...
+        "description": "returns preview stack template for given template and custom variables",
+        "responses": {
+          "200": {
+            "description": "stacktemplate preview in provided format",
+            "schema": {
+              "type": "object",
+              "properties": {
+                "template": {
+                  "type": "string",
+                  "default": "Hello gokmen -- bar"
```
Comment (sinan) [inline website/swagger.json]: intentional? 😏

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 6]
```diff
@@ -400,6 +402,55 @@ module.exports = class JStackTemplate extends Module
...
+  #   }
+  #
+  # @example return
+  #
+  #   {
+  #     "template": "Hello gokmen -- bar",
+  #     "errors": {},
+  #     "warnings": {}
+  #   }
+  #
```
Comment (sinan) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: block comment would be better for usage description, unless we do this on other places.

**Review by sinan — APPROVED** [review]
I have some nice to haves, only important one is the regex comment.

File: website/swagger.json (hunk 5 above)
Comment (gokmen) [inline website/swagger.json]: :) yes, it's going to show an output with the logged in user, so I've tried with my own and pasted the output.

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee (hunk 6 above)
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: 👍

**Pull request merged**